src/proxy/http/metrics/service.rs

```diff
@@ -262,16 +263,20 @@ where
     F: Future<Item = http::Response<B>>,
     B: Payload,
     C: ClassifyResponse<Error = h2::Error> + Send + Sync + 'static,
-    C::Class: Hash + Eq,
+    C::Class: Hash + Eq + Send + Sync,
 {
     type Item = http::Response<ResponseBody<B, C::ClassifyEos>>;
     type Error = F::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let rsp = try_ready!(self.inner.poll());
+        let mut rsp = try_ready!(self.inner.poll());
 
         let classify = self.classify.take().map(|c| c.start(&rsp));
 
+        if let Some(class) = classify.as_ref().and_then(|c| c.no_body()) {
+            rsp.extensions_mut().insert(class);
+        }
```

Comment:
OK, if I understand this properly, the retry layer depends on the metric layer setting this classification?

Can these things be decoupled?  This is subtle, and probably fragile.

Reply:
Specifically, if a response fails mid-stream, we'd mark it as a failure in metrics, though it wouldn't be retryable. What are the disadvantages of duplicating the classification logic?  I imagine the CPU overhead is infinitesimal...  Should the `classify` trait perhaps be specific to metrics? I imagine a retry-specific classification API would be a lot simpler...

Reply:
This part was added so that a "default" classifier could be used, if the route doesn't include any response classifiers. An alternative is to duplicate the default classifier stuff in retry as well.

Reply:
I don't think this is a blocker, just talking through it a bit...

I'm fine with leaving this as-is if we add a little more documentation that covers:
- the role of no_body in the classify API
- the soft extension-coupling between metrics and retries